python/modules/IcePy/Communicator.cpp

```diff
@@ -289,6 +282,65 @@ communicatorDestroy(CommunicatorObject* self, PyObject* /*args*/)
     }
 }
 
+extern "C" PyObject*
+communicatorDestroyAsync(CommunicatorObject* self, PyObject* /*args*/)
+{
+    assert(self->communicator);
+
+    auto vfm = dynamic_pointer_cast<ValueFactoryManager>((*self->communicator)->getValueFactoryManager());
+    assert(vfm);
+
+    auto futureType = reinterpret_cast<PyTypeObject*>(lookupType("Ice.Future"));
+    assert(futureType);
+
+    PyObjectHandle emptyArgs{PyTuple_New(0)};
+    PyObjectHandle future{futureType->tp_new(futureType, emptyArgs.get(), nullptr)};
+    if (!future.get())
+    {
+        return nullptr;
+    }
+
+    // Call Ice.Future.__init__
+    futureType->tp_init(future.get(), emptyArgs.get(), nullptr);
+
+    // Create a new reference to prevent premature release of the future object. The reference will be released by
+    // the destroy callback.
+    PyObject* futureObject = Py_NewRef(future.get());
+
+    (*self->communicator)
+        ->destroyAsync(
+            [self, vfm, futureObject]()
+            {
+                // Ensure the current thread is able to call into Python.
+                AdoptThread adoptThread;
```

Comment:
Just to confirm: AdoptThread is recursive - it works fine even if you have already the GIL locked?